connector/connector_ldap.go

```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
 }
 
 func (c *LDAPConnector) Sync() chan struct{} {
+	go func() {
+		for {
+			alive, killed := c.idp.ldapPool.Ping()
+			if alive > 0 {
+				log.Infof("Connector ID=%v idle_conns=%v", c.id, alive)
+			}
+			if killed > 0 {
+				log.Warningf("Connector ID=%v closed %v dead connections.", c.id, killed)
+			}
+			time.Sleep(c.idp.ldapPool.PoolCheckTimer)
+		}
+	}()
 	return make(chan struct{})
 }
 
 func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+// A LDAPPool is a Connection Pool for LDAP connections
+// Initialize exported fields and use Get() to get a connection.
+// Use Put() to put it back into the pool.
+type LDAPPool struct {
+	m              sync.Mutex
+	conns          map[*ldap.Conn]struct{}
+	MaxIdleConn    int
+	PoolCheckTimer time.Duration
+	ServerHost     string
+	ServerPort     uint16
+	UseTLS         bool
+	UseSSL         bool
+	TlsConfig      *tls.Config
+}
+
+// Get returns a random connection from the pool. A new connection is returned if there are
+// no connections available in the pool.
+func (p *LDAPPool) Get() (*ldap.Conn, error) {
+	p.m.Lock()
+	for conn := range p.conns {
+		delete(p.conns, conn)
+		p.m.Unlock()
+		return conn, nil
+	}
+	p.m.Unlock()
+	return p.ldapConnect()
+}
+
+// Put makes a connection ready for re-use and puts it back into the pool. If the connection
+// cannot be reused it is discarded. If there allready are MaxIdleConn connections in the pool
+// the connection is discarded.
+func (p *LDAPPool) Put(c *ldap.Conn) {
+	p.m.Lock()
+	if p.conns == nil {
+		// First call to Put, initialize map
+		p.conns = make(map[*ldap.Conn]struct{})
+	}
+	if len(p.conns)+1 > p.MaxIdleConn {
+		p.m.Unlock()
+		c.Close()
+		return
+	}
+	p.m.Unlock()
+	// drop to anonymous bind
+	err := c.Bind("", "")
+	if err != nil {
+		// unsupported or disallowed, throw away connection
+		log.Warningf("Unable to re-use LDAP Connection after failure to bind anonymously: %v", err)
+		c.Close()
+		return
+	}
+	p.m.Lock()
+	p.conns[c] = struct{}{}
+	p.m.Unlock()
+}
+
+// getConn attempts to remove the provided connection from the pool. If getConn returns false
+// another routine is using the connection and the caller should discard the pointer.
+func (p *LDAPPool) getConn(conn *ldap.Conn) bool {
```

Comment:
I do see your point, but It feels weird to use "remove" to get a connection to do some work with. But I'll change it, I may try to come up with a alternative name that encompasses both concerns.